In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [3]:
import pandas as pd
import numpy as np
import os

rf_file = "../datasets/processed/day19_random_forest_results.csv"

print("File exists:", os.path.exists(rf_file))

rf_results = pd.read_csv(rf_file)

print(rf_results)

File exists: True
           Model       MAE         MSE       RMSE        R2
0  Random Forest  15.46964  511.864893  22.624431 -0.067891


In [4]:
# Linear Regression results
linear_results = pd.DataFrame({
    "Model": ["Linear Regression"],
    "MAE": [12.896688748879361],
    "MSE": [303.6465688580839],
    "RMSE": [17.425457493508855],
    "R2": [0.3665095339776454]
})

# Random Forest results
rf_results = pd.read_csv(
    "../datasets/processed/day19_random_forest_results.csv"
)

# XGBoost results
xgb_results = pd.DataFrame({
    "Model": ["XGBoost"],
    "MAE": [15.1574],
    "MSE": [403.3640],
    "RMSE": [20.0839],
    "R2": [0.1585]
})

# Combine all models
comparison = pd.concat(
    [
        linear_results,
        rf_results[["Model", "MAE", "MSE", "RMSE", "R2"]],
        xgb_results
    ],
    ignore_index=True
)

print(comparison)

               Model        MAE         MSE       RMSE        R2
0  Linear Regression  12.896689  303.646569  17.425457  0.366510
1      Random Forest  15.469640  511.864893  22.624431 -0.067891
2            XGBoost  15.157400  403.364000  20.083900  0.158500


In [5]:
comparison.round(4)

,Model,MAE,MSE,RMSE,R2
0,Linear Regression,12.8967,303.6466,17.4255,0.3665
1,Random Forest,15.4696,511.8649,22.6244,-0.0679
2,XGBoost,15.1574,403.3640,20.0839,0.1585


In [6]:
best_mae = comparison.loc[
    comparison["MAE"].idxmin(), "Model"
]

best_rmse = comparison.loc[
    comparison["RMSE"].idxmin(), "Model"
]

best_r2 = comparison.loc[
    comparison["R2"].idxmax(), "Model"
]

print("Best model based on MAE :", best_mae)
print("Best model based on RMSE:", best_rmse)
print("Best model based on R²  :", best_r2)

Best model based on MAE : Linear Regression
Best model based on RMSE: Linear Regression
Best model based on R²  : Linear Regression


In [7]:
comparison.to_csv(
    "../datasets/processed/day21_model_comparison.csv",
    index=False
)

print("Day 21 comparison saved successfully!")

Day 21 comparison saved successfully!


In [8]:
print(comparison.round(4))

               Model      MAE       MSE     RMSE      R2
0  Linear Regression  12.8967  303.6466  17.4255  0.3665
1      Random Forest  15.4696  511.8649  22.6244 -0.0679
2            XGBoost  15.1574  403.3640  20.0839  0.1585


In [9]:
comparison.to_csv(
    "../datasets/processed/day21_model_comparison.csv",
    index=False
)

print("Model comparison saved successfully!")

Model comparison saved successfully!


In [10]:
print(comparison)

               Model        MAE         MSE       RMSE        R2
0  Linear Regression  12.896689  303.646569  17.425457  0.366510
1      Random Forest  15.469640  511.864893  22.624431 -0.067891
2            XGBoost  15.157400  403.364000  20.083900  0.158500


## Final Model Selection

Based on the evaluation results, Linear Regression was selected as the best-performing model.

### Final Results

| Model | MAE | RMSE | R² |
|---|---:|---:|---:|
| Linear Regression | 12.8967 | 17.4255 | 0.3665 |
| Random Forest | 15.4696 | 22.6244 | -0.0679 |
| XGBoost | 15.1574 | 20.0839 | 0.1585 |

Linear Regression achieved the lowest MAE and RMSE and the highest R² score.

Therefore, Linear Regression will be used as the primary predictive model for the subsequent Dividend Puzzle theory analysis.

The results also show that the more complex tree-based models did not outperform the simpler Linear Regression model on the current dataset.